### ISSUE TITLE
cmake build fails from release tgz

### ISSUE THREAD

**Issue by czanik** [syslog-ng/syslog-ng#4779]
I tried a cmake build on MacOS. While it worked when I used it directly on checked out git sources, trying to build syslog-ng using a tgz built from the same git snapshot, the build fails in the configuration phase. One problem is the freshly introduced gperf (https://github.com/syslog-ng/syslog-ng/pull/4763), the other one is Python, which is especially interesting, as all Python support is turned off:

```
bash-3.2$ cmake --install-prefix    $PWD/build/install      -B build .  -DCMAKE_BUILD_TYPE=Debug  -DBUILD_TESTING=OFF -DIVYKIS_SOURCE=system -DPYTHON_VERSION=3 -DENABLE_AFSMTP=OFF  -DENABLE_SNMP=OFF  -DENABLE_JAVA=OFF -DENABLE_JAVA_MODULES=OFF -DENABLE_PYTHON=OFF -DENABLE_PYTHON_MODULES=OFF -DENABLE_MQTT=OFF -DENABLE_PACCT=OFF -DENABLE_CPP=OFF -DENABLE_EXAMPLE_MODULES=OFF   -DSUMMARY_VERBOSE=ON --fresh       -DENABLE_BIGQUERY=OFF
-- The C compiler identification is AppleClang 15.0.0.15000100
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:43 (include):
  include could not find requested file:

    FindGperf


-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of struct sockaddr_storage
-- Check size of struct sockaddr_storage - done
-- Check size of struct ucred
-- Check size of struct ucred - failed
-- Check size of struct cmsgcred
-- Check size of struct cmsgcred - failed
-- Performing Test SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR
-- Performing Test SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR - Success
-- Check size of struct sockaddr_in6
-- Check size of struct sockaddr_in6 - done
-- Looking for TCP_KEEPIDLE
-- Looking for TCP_KEEPIDLE - not found
-- Looking for TCP_KEEPINTVL
-- Looking for TCP_KEEPINTVL - found
-- Looking for TCP_KEEPCNT
-- Looking for TCP_KEEPCNT - found
-- Check size of O_LARGEFILE
-- Check size of O_LARGEFILE - failed
-- Performing Test SYSLOG_NG_HAVE_THREAD_KEYWORD
-- Performing Test SYSLOG_NG_HAVE_THREAD_KEYWORD - Success
-- Looking for strtoll
-- Looking for strtoll - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for getline
-- Looking for getline - found
-- Looking for strtok_r
-- Looking for strtok_r - found
-- Looking for strtoimax
-- Looking for strtoimax - found
-- Looking for inet_aton
-- Looking for inet_aton - found
-- Looking for inet_ntoa
-- Looking for inet_ntoa - found
-- Looking for getutent
-- Looking for getutent - not found
-- Looking for getutxent
-- Looking for getutxent - found
-- Looking for getaddrinfo
-- Looking for getaddrinfo - found
-- Looking for getnameinfo
-- Looking for getnameinfo - found
-- Looking for getprotobynumber_r
-- Looking for getprotobynumber_r - not found
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for gmtime_r
-- Looking for gmtime_r - found
-- Looking for localtime_r
-- Looking for localtime_r - found
-- Looking for getrandom
-- Looking for getrandom - not found
-- Looking for fmemopen
-- Looking for fmemopen - found
-- Looking for memrchr
-- Looking for memrchr - not found
-- Looking for strcasestr
-- Looking for strcasestr - found
-- Looking for pread
-- Looking for pread - found
-- Looking for pwrite
-- Looking for pwrite - found
-- Looking for posix_fallocate
-- Looking for posix_fallocate - not found
-- Looking for timezone
-- Looking for timezone - found
-- Looking for include file utmp.h
-- Looking for include file utmp.h - found
-- Looking for include file utmpx.h
-- Looking for include file utmpx.h - found
-- Looking for include file dlfcn.h
-- Looking for include file dlfcn.h - found
-- Looking for include file getopt.h
-- Looking for include file getopt.h - found
-- Performing Test UTMPX_HAS_UT_TYPE
-- Performing Test UTMPX_HAS_UT_TYPE - Success
-- Performing Test UTMP_HAS_UT_TYPE
-- Performing Test UTMP_HAS_UT_TYPE - Failed
-- Performing Test UTMPX_HAS_UT_USER
-- Performing Test UTMPX_HAS_UT_USER - Success
-- Performing Test UTMP_HAS_UT_USER
-- Performing Test UTMP_HAS_UT_USER - Failed
-- Performing Test SYSLOG_NG_HAVE_STRUCT_TM_TM_GMTOFF
-- Performing Test SYSLOG_NG_HAVE_STRUCT_TM_TM_GMTOFF - Success
-- Looking for SO_MEMINFO
-- Looking for SO_MEMINFO - not found
-- Looking for linux/sock_diag.h
-- Looking for linux/sock_diag.h - not found
-- Found BISON: /usr/local/opt/bison/bin/bison (found suitable version "3.8.2", minimum required is "3.7.6") 
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2") 
-- Found GLIB: /usr/local/Cellar/glib/2.78.3/include/glib-2.0;/usr/local/Cellar/glib/2.78.3/lib/glib-2.0/include (found suitable version "2.78.3", minimum required is "2.32.0") 
-- Looking for g_list_copy_deep
-- Looking for g_list_copy_deep - found
-- Looking for g_ptr_array_find_with_equal_func
-- Looking for g_ptr_array_find_with_equal_func - found
-- Looking for g_canonicalize_filename
-- Looking for g_canonicalize_filename - found
CMake Warning at cmake/Modules/LibFindMacros.cmake:299 (message):
  WARNING: MISSING PACKAGE

  We could not find development headers for RabbitMQ.  Do you have the
  necessary dev package installed? This package is NOT REQUIRED and you may
  ignore this warning but by doing so you may miss some functionality of
  syslog-ng.

  Relevant CMake configuration variables:

    RabbitMQ_INCLUDE_DIR=<not found>
    RabbitMQ_LIBRARY=<not found>

  You may use CMake GUI, cmake -D or ccmake to modify the values.  Delete
  CMakeCache.txt to discard all values and force full re-detection if
  necessary.

Call Stack (most recent call first):
  cmake/Modules/FindRabbitMQ.cmake:29 (libfind_process)
  CMakeLists.txt:243 (find_package)


-- Found IVYKIS 
-- Looking for iv_work_pool_submit_continuation
-- Looking for iv_work_pool_submit_continuation - not found
-- Found OpenSSL: /usr/local/Cellar/openssl@3/3.2.0_1/lib/libcrypto.dylib (found version "3.2.0")  
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
CMake Warning (dev) at cmake/Modules/FindLIBNET.cmake:44 (EXEC_PROGRAM):
  Policy CMP0153 is not set: The exec_program command should not be called.
  Run "cmake --help-policy CMP0153" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Use execute_process() instead.
Call Stack (most recent call first):
  CMakeLists.txt:259 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/Modules/FindLIBNET.cmake:45 (EXEC_PROGRAM):
  Policy CMP0153 is not set: The exec_program command should not be called.
  Run "cmake --help-policy CMP0153" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Use execute_process() instead.
Call Stack (most recent call first):
  CMakeLists.txt:259 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found LIBNET: -L/usr/local/Cellar/libnet/1.3/lib -lnet  
-- Could NOT find WRAP (missing: WRAP_LIBRARY WRAP_INCLUDE_DIR) 
-- Found CRITERION 
-- Checking for one of the modules 'libsystemd>=245'
-- Found lib version 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
CMake Warning (dev) at CMakeLists.txt:306 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonInterp: /usr/local/bin/python3 (found suitable exact version "3.11.7") 
CMake Warning (dev) at CMakeLists.txt:307 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonLibs: /usr/local/Frameworks/Python.framework/Versions/3.11/lib/libpython3.11.dylib (found suitable exact version "3.11.7") 
-- Detected pythonlib version: 3.11.7
CMake Error at CMakeLists.txt:311 (include):
  include could not find requested file:

    python_build_venv


-- Looking for EVP_MD_CTX_reset
-- Looking for EVP_MD_CTX_reset - found
-- Looking for ASN1_STRING_get0_data
-- Looking for ASN1_STRING_get0_data - found
-- Looking for SSL_CTX_get0_param
-- Looking for SSL_CTX_get0_param - found
-- Looking for X509_STORE_CTX_get0_cert
-- Looking for X509_STORE_CTX_get0_cert - found
-- Looking for X509_get_extension_flags
-- Looking for X509_get_extension_flags - found
-- Looking for DH_set0_pqg
-- Looking for DH_set0_pqg - found
-- Looking for BN_get_rfc3526_prime_2048
-- Looking for BN_get_rfc3526_prime_2048 - found
-- Looking for SSL_CONF_CTX_new
-- Looking for SSL_CONF_CTX_new - found
-- Looking for SSL_CTX_set_num_tickets
-- Looking for SSL_CTX_set_num_tickets - found
-- Looking for SSL_CTX_set_ciphersuites
-- Looking for SSL_CTX_set_ciphersuites - found
-- Looking for SSL_CTX_set1_sigalgs_list
-- Looking for SSL_CTX_set1_sigalgs_list - found
-- Looking for SSL_CTX_set1_client_sigalgs_list
-- Looking for SSL_CTX_set1_client_sigalgs_list - found
-- Looking for SSL_CTX_set_min_proto_version
-- Looking for SSL_CTX_set_min_proto_version - found
-- Looking for SSL_add_file_cert_subjects_to_stack
-- Looking for SSL_add_file_cert_subjects_to_stack - found
-- Looking for SSL_add_dir_cert_subjects_to_stack
-- Looking for SSL_add_dir_cert_subjects_to_stack - found
-- Checking for module 'libpcre2-8'
--   Found libpcre2-8, version 10.42
CMake Error at lib/CMakeLists.txt:384 (gperf_generate):
  Unknown CMake command "gperf_generate".


-- Configuring incomplete, errors occurred!

```

**Comment by Schamschula** [comment]
I just ran into this same issues while attempting to bring `syslog-ng` up to date for MacPorts.

I managed to work around the `FindGperf` issue by installing the `kde-extra-cmake-modules` port, and `Unknown CMake command "gperf_generate"` by patching `lib/CMakeLists.txt` to change `gperf_generate` to `ecm_gperf_generate`

However, to deal with
```
include could not find requested file:

    python_build_venv
```
I tried commenting out `include(python_build_venv)`. This allowed cmake to finish cleanly.

While building I ran into a new issue:

```
:info:build [ 81%] Building C object modules/afsnmp/CMakeFiles/afsnmp.dir/afsnmpdest.c.o
:info:build [ 81%] Built target stardate
:info:build /Library/Developer/CommandLineTools/usr/bin/make  -f modules/tagsparser/CMakeFiles/tags-parser.dir/build.make modules/tagsparser/CMakeFiles/tags-parser.dir/depend
:info:build cd /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/modules/afsnmp && /usr/bin/clang -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_68 -D_GNU_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -D__APPLE_USE_RFC_3542 -Dafsnmp_EXPORTS -I/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build -I/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/modules/afsnmp -I/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/modules/afsnmp -I/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/lib -I/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/lib -isystem /opt/local/include -isystem /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/ivykis-install/include -isystem /opt/local/include/json-c -isystem /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/lib/eventlog/src -isystem /opt/local/include/glib-2.0 -isystem /opt/local/lib/glib-2.0/include -isystem /opt/local/libexec/openssl3/include -pipe -Os -DNDEBUG -I/opt/local/include -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -std=gnu99 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -mmacosx-version-min=14.0 -fPIC -DNETSNMP_ENABLE_IPV6 -fno-strict-aliasing -g -Os -pipe -DBUILD=177 -DMACOSX_DEPLOYMENT_TARGET=14.5 --sysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -iframeworkwithsysroot /System/Library/Frameworks/-f985-11ee-b563-fe8bc7981bff/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.Internal.sdk/System/Library/PrivateFrameworks/ -f985-11ee-b563-fe8bc7981bff/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.Internal.sdk/System/Library/Frameworks/ -iwithsysroot /usr/local/libressl/include -L/AppleInternal/Library/BuildRoots/91a344b1-f985-11ee-b563-fe8bc7981bff/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.Internal.sdk/usr/local/libressl/lib -Udarwin23 -Ddarwin23=darwin23 -I. -I/usr/local/include -Wshadow -Wno-stack-protector -Wno-unused-parameter -Wno-variadic-macros -Wimplicit-function-declaration -Wnested-externs -Wstrict-prototypes -Wswitch-default -Wimplicit-int -Wall -Wuninitialized -Wdeprecated -Wdeprecated-declarations -Woverflow -Wdouble-promotion -Wfloat-equal -Wpointer-arith -Wpointer-sign -Wmissing-format-attribute -Wold-style-definition -Wundef -Wignored-qualifiers -Wfloat-conversion -Wbad-function-cast -MD -MT modules/afsnmp/CMakeFiles/afsnmp.dir/afsnmpdest.c.o -MF CMakeFiles/afsnmp.dir/afsnmpdest.c.o.d -o CMakeFiles/afsnmp.dir/afsnmpdest.c.o -c /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/modules/afsnmp/afsnmpdest.c
:info:build make[2]: Entering directory `/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build'
:info:build make[2]: Leaving directory `/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build'
:info:build /Library/Developer/CommandLineTools/usr/bin/make  -f modules/system-source/CMakeFiles/system-source.dir/build.make modules/system-source/CMakeFiles/system-source.dir/build
:info:build make[2]: Entering directory `/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build'
:info:build [ 81%] Generating tags-parser-grammar.y
:info:build cd /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/modules/tagsparser && /opt/local/bin/python3 /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/lib/merge-grammar.py /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/modules/tagsparser/tags-parser-grammar.ym > /opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build/modules/tagsparser/tags-parser-grammar.y
:info:build clang: error: unknown argument: '-f985-11ee-b563-fe8bc7981bff/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.Internal.sdk/System/Library/Frameworks/'
:info:build make[2]: *** [modules/afsnmp/CMakeFiles/afsnmp.dir/afsnmpdest.c.o] Error 1
:info:build make[2]: Leaving directory `/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/build'
:info:build make[1]: *** [modules/afsnmp/CMakeFiles/afsnmp.dir/all] Error 2

```

Note: I also set `-DENABLE_AFSMTP=OFF`, which was ignored, just like `-DENABLE_PYTHON=OFF`

**Comment by Schamschula** [comment]
The above error can be worked around by setting `-DENABLE_AFSNMP=OFF`. However, `Criterion` seems not to be an optional package as the build now fails with 

```
/opt/local/var/macports/build/_Users_marius_Development_MacPorts_ports_sysutils_syslog-ng/syslog-ng/work/syslog-ng-4.7.1/libtest/msg_parse_lib.c:28:10: fatal error: 'criterion/criterion.h' file not found
#include <criterion/criterion.h>
         ^~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [libtest/CMakeFiles/libtest.dir/msg_parse_lib.c.o] Error 1
```
 even though `cmake` couldn't find `Criterion` and I had set `-DBUILD_TESTING=OFF`

**Comment by HofiOne** [comment]
AFSNMP should work now, please check the brew packages you need and how to set up the env [here](https://syslog-ng.github.io/dev-guide/chapter_4/section_2/README), plz drop me a line if it still does not work for you

Unfortunately, criterion is still hardly bound to the cmake builds in some cases, and as you experienced can be turned off only via `-DBUILD_TESTING=OFF`
It is on my list, I'd like to address this issue in the near future.

**Comment by HofiOne** [comment]
> I'm not using Homebrew, but MacPorts. The Homebrew formula still uses autotools. However, there is a hard error in the build `ld: library not found for -lSystem` which you get to after dealing with `ld: unknown options: -noall_load` for current clang compilers. I'm building under macOS 14.5 23F79 arm64 and Xcode 15.4 15F31d with Apple clang version 15.0.0 (clang-1500.3.9.4).

It's been a while since I tried it with MacPorst, but it surely must work with Homebrew as the CI builds are using the same, both on macOS 13, 14 - x86_64, and ARM64 (also, I'm on macOS 14.5, M1 Max Arm64, Apple clang version 15.0.0 (clang-1500.3.9.4))
I suggest you give it a try with Homebrew.


> My current WIP Portfile [Schamschula/macports@8bd779b](https://github.com/Schamschula/macports/commit/8bd779b834df1b9f3273e7b4f67a320b60420378) passes `-DBUILD_TESTING=OFF` to cmake, but that doesn't prevent the build error, as I mentioned above. Only after installing `criterion` did I get a clean build.

you are right, the issue is that it CANNOT be turned off via that option, that was the problem I wanted to fix, and I can remember now.

**Comment by Schamschula** [comment]
Speaking of things that don't get turned off: `-DENABLE_MQTT=OFF` and `-DENABLE_KAFKA=OFF` don't do anything either.

**Comment by HofiOne** [comment]
> Speaking of things that don't get turned off: `-DENABLE_MQTT=OFF` and `-DENABLE_KAFKA=OFF` don't do anything either.

I will double-check these, but it is very strange, I was using them multiple times earlier without any issues.

**Comment by Schamschula** [comment]
In both cases the tests failed on my machine and the modules were turned off by cmake, but could have lead to an opportunistic inclusion of a module, if the library had been present.

**Comment by Schamschula** [comment]
Both issues are still there, I just worked around them:

1) there is no cmake `FindGperf` unless you install `kde-extra-cmake-modules`. The module needs to be added to `cmake/Modules`
2) I commented out a line in `CMakeLists.txt` to deal with the Python issue. There needs to be a better solution.

**Comment by HofiOne** [comment]
`FIndGperf` must be at `${PROJECT_ROOT}/cmake/Modules/FindGperf.cmake`, `python_build_venv` must be created by the `${PROJECT_ROOT}/cmake/python_build_venv.cmake` normally

probably these are missing somehow from the tar.gz pack, I will double-check it.

**Comment by HofiOne** [comment]
I've fixed these, but with the latest 4.8.0 tar.gz there are multiple other issues.
I will try to fix them all and will submit together with the fixes of these.

**Comment by HofiOne** [comment]
> In both cases the tests failed on my machine and the modules were turned off by cmake, but could have lead to an opportunistic inclusion of a module, if the library had been present.

@Schamschula could you please add details of the referred `opportunistic inclusion` of a given module to this issue #5034 (I checked these switches and could not find any similar issues)

TIA!

**Comment by Schamschula** [comment]
Under MacPorts we refer to `opportunistic inclusion` when a package/library is installed on the computer and is found by the configure script, even though the Portfile does not specifically call it out as a dependency. This can cause unexpected build and runtime errors, as these are untested by the port maintainer.

In this case the modules in question are MQTT and Kafka. Both exist under MacPorts, but may be missing parts, e.g. cmake configuration files, preventing cmake from detecting these libraries. This may change as the ports get updated in the future.

**Comment by HofiOne** [comment]
Thanks for your explanation.

When building with CMake, the module check should follow this flow (using ENABLE_MQTT as an example):

1. If the module switch is not presented (not set from the configure command line) or it is set to ON (ENABLE_MQTT=ON), the dependency package CMake discovery runs (find_package, etc.).
2. The module switch is either left untouched, if it was already defined from the command line, or it is set depending on the package discovery result: ON if the package is found, OFF if not found.
3. If the module switch is presented (set from the configure command line) and is set to ON (ENABLE_MQTT=ON), but the dependency package could not be found, an error is displayed, and the configure step is stopped.
4. As a final step, if the resulting value of the module switch is OFF (ENABLE_MQTT=OFF), the module is skipped.
5. Otherwise, all the remaining required CMake configuration steps take effect.

Unfortunately, not all module CMake files follow these rules yet.

This means that you must explicitly DISABLE the given module to prevent the implicit search for its dependencies.

The strange thing is that KAFKA more or less followed the same rules, so you should not have seen the mentioned `opportunistic inclusion`, I guess.

I modified both CMake files to behave the same way (and will try to consolidate all of them later in #5049), but currently, this is all I can do to address this issue. Sorry.

**Comment by barracuda156** [comment]
@Schamschula What is even the need to search for `gperf`? It is just an executable, it should be in the PATH.

**Comment by Schamschula** [comment]
> @Schamschula What is even the need to search for `gperf`? It is just an executable, it should be in the PATH.

Except that cmake is looking for `gperf_generate` which isn't part of the `gperf`  package, but rather part of the `cmake` infrastructure. However, it is not part of default `cmake` but is provided by the `kde-extra-cmake-modules` port, see comment above.

**Comment by barracuda156** [comment]
@Schamschula Got it. Anyway, all these issues are fixed in the master (but not back-ported to the release).

I have made some fixes and improvements today, minor changes left before PRs are merged. I can make then a `-devel` subport so that in can be tested prior to updating the main one.

**Comment by Schamschula** [comment]
> @Schamschula Got it. Anyway, all these issues are fixed in the master (but not back-ported to the release).
> 
> I have made some fixes and improvements today, minor changes left before PRs are merged. I can make then a `-devel` subport so that it can be tested prior to updating the main one. (Also to update the main one we need to wait for the new release, otherwise too much to patch.)

Also see: https://github.com/Schamschula/macports/tree/master/sysutils/syslog-ng

**Comment by barracuda156** [comment]
> > @Schamschula Got it. Anyway, all these issues are fixed in the master (but not back-ported to the release).
> > I have made some fixes and improvements today, minor changes left before PRs are merged. I can make then a `-devel` subport so that it can be tested prior to updating the main one. (Also to update the main one we need to wait for the new release, otherwise too much to patch.)
> 
> Also see: https://github.com/Schamschula/macports/tree/master/sysutils/syslog-ng

@Schamschula Oh, awesome, I will borrow destroot code from you, since I haven’t sorted that part yet.

One question: is there some reason to prefer patching clock_gettime instead of just using `legacysupport`? I did the latter, it works. (Then also MAP_ANON patch becomes redundant, though ideally it should rather be upstreamed.)

**Comment by HofiOne** [comment]
> Also see: https://github.com/Schamschula/macports/tree/master/sysutils/syslog-ng

@Schamschula I think most of these patches can easily be part of the syslog-ng/master.
I guess you are maintaining the macPorts version of syslog-ng, am I right?
If you can create an issue after each syslog-ng release about the issues you faced, I will address them ASAP, (also, any PRs are also kindly welcome, of course ;) )

**Comment by barracuda156** [comment]
@HofiOne The port has no maintainer at the moment (would you like to be one?), just @Schamschula and myself independently decided to finally update it.
I wanted to do that together with fixing here https://github.com/macports/macports-ports/pull/25715 but bumped into issues which we discussed earlier.

**Comment by barracuda156** [comment]
@Schamschula I remembered now: I wanted to add `criterion` before (not sure for which dependent), but it is broken on powerpc due to boxfort:
```
--->  Configuring criterion
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2" && /opt/local/bin/meson setup --prefix=/opt/local /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2 /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build --wrap-mode=default 
Cloning into 'debugbreak'...
HEAD is now at 83bf7e9 Use __builtin_debugtrap wherever available
Cloning into 'klib'...
HEAD is now at cdb7e92 Merge pull request #62 from innoink/master
Cloning into 'nanomsg'...
Cloning into 'nanopb'...
HEAD is now at b97aa65 Publishing nanopb-0.4.7
Cloning into 'boxfort'...
HEAD is now at c91d718 Add riscv64 support
The Meson build system
Version: 1.5.1
Source dir: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2
Build dir: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build
Build type: native build
Project name: criterion
Project version: 2.4.2
C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
Host machine cpu family: ppc
Host machine cpu: power macintosh
Program python3 found: YES
Program git found: YES (/opt/local/bin/git)
meson.build:39: WARNING: add_languages is missing native:, assuming languages are wanted for both host and build.
C++ compiler for the host machine: ccache /usr/bin/g++-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-g++-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
C++ linker for the host machine: /usr/bin/g++-4.2 ld64 87.3.1
Compiler for C supports arguments -Wno-unused-parameter: YES 
Compiler for C supports arguments -Wno-unused-value: YES 
Compiler for C supports arguments -fvisibility=hidden: YES 
Compiler for C supports arguments -fexceptions: YES 
Compiler for C supports arguments /SAFESEH:NO: NO 
Compiler for C supports arguments /source-charset:utf-8: NO 
Checking for function "clock_gettime" : NO 
Checking for function "fopencookie" : NO 
Checking for function "funopen" : NO 
Checking for function "getcwd" : NO 
Checking for function "isatty" : NO 
Checking for function "nl_langinfo" : NO 
Checking for function "open_memstream" : NO 
Checking for function "strtok_r" : NO 
Checking for function "strtok_s" : NO 
Header "time.h" has symbol "CLOCK_MONOTONIC_RAW" : NO 
Checking for function "GetCurrentDirectory" : NO 
Checking for function "PathIsRelative" : NO 
Has header "synchapi.h" : NO 
Program gettext found: YES (/opt/local/bin/gettext)
Program msgfmt found: YES (/opt/local/bin/msgfmt)
Program msginit found: YES (/opt/local/bin/msginit)
Program msgmerge found: YES (/opt/local/bin/msgmerge)
Program xgettext found: YES (/opt/local/bin/xgettext)
Found pkg-config: YES (/opt/local/bin/pkg-config) 0.29.2
Found CMake: /opt/local/bin/cmake (3.30.3)
Run-time dependency debugbreak found: NO (tried pkgconfig, framework and cmake)
Looking for a fallback subproject for the dependency debugbreak

Executing subproject debugbreak 

debugbreak| Project name: debugbreak
debugbreak| Project version: undefined
debugbreak| C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
debugbreak| C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
debugbreak| Build targets in project: 5
debugbreak| Subproject debugbreak finished.

Dependency debugbreak from subproject subprojects/debugbreak found: YES undefined
Run-time dependency klib found: NO (tried pkgconfig, framework and cmake)
Looking for a fallback subproject for the dependency klib

Executing subproject klib 

klib| Project name: klib
klib| Project version: undefined
klib| C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
klib| C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
klib| Build targets in project: 5
klib| Subproject klib finished.

Dependency klib from subproject subprojects/klib found: YES undefined
Run-time dependency threads found: YES
Run-time dependency nanomsg found: NO (tried pkgconfig, framework and cmake)

Executing subproject nanomsg-cmake method cmake 

nanomsg-cmake| Found CMake: /opt/local/bin/cmake (3.30.3)

| Configuring the build directory with CMake version 3.30.3
| Running CMake with: -G Ninja -DCMAKE_INSTALL_PREFIX=/opt/local -DNN_TESTS=False -DNN_TOOLS=False -DNN_STATIC_LIB=True -DBUILD_SHARED_LIBS=False -DCMAKE_POSITION_INDEPENDENT_CODE=True -DCMAKE_BUILD_TYPE=Debug
| - build directory:          /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanomsg/__CMake_build
| - source directory:         /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2/subprojects/nanomsg
| - toolchain file:           /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanomsg/__CMake_build/CMakeMesonToolchainFile.cmake
| - preload file:             /opt/local/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/mesonbuild/cmake/data/preload.cmake
| - trace args:               --trace-expand --trace-format=json-v1 --no-warn-unused-cli --trace-redirect=cmake_trace.txt
| - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089, CMP0102]

| CMake Deprecation Warning at CMakeLists.txt:27 (cmake_minimum_required):
| Compatibility with CMake < 3.5 will be removed from a future version of
| CMake.

| Update the VERSION argument <min> value or use a ...<max> suffix to tell
| CMake that the project does not need compatibility with older versions.

| Put cmake in trace mode, but with variables expanded.
| Put cmake in trace mode and sets the trace output format.
| Not searching for unused variables given on the command line.
| Put cmake in trace mode and redirect trace output to a file instead of stderr.
| Trace will be written to cmake_trace.txt

| -- The C compiler identification is GNU 4.2.1
| -- Checking whether C compiler has -isysroot
| -- Checking whether C compiler has -isysroot - yes
| -- Checking whether C compiler supports OSX deployment target flag
| -- Checking whether C compiler supports OSX deployment target flag - yes
| -- Detecting C compiler ABI info
| -- Detecting C compiler ABI info - done
| -- Check for working C compiler: /usr/bin/gcc-4.2 - skipped
| -- Detecting C compile features
| -- Detecting C compile features - done
| -- Detected nanomsg ABI v6 (v6.0.1)
| -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
| -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
| -- Found Threads: TRUE
| -- OS System is Darwin
| -- OS Version is 10.0.0d2
| -- Looking for gethrtime
| -- Looking for gethrtime - not found
| -- Looking for socketpair
| -- Looking for socketpair - found
| -- Looking for eventfd
| -- Looking for eventfd - not found
| -- Looking for pipe
| -- Looking for pipe - found
| -- Looking for pipe2
| -- Looking for pipe2 - not found
| -- Looking for accept4
| -- Looking for accept4 - not found
| -- Looking for epoll_create
| -- Looking for epoll_create - not found
| -- Looking for kqueue
| -- Looking for kqueue - found
| -- Looking for poll
| -- Looking for poll - found
| -- Looking for getaddrinfo_a in anl
| -- Looking for getaddrinfo_a in anl - not found
| -- Looking for clock_gettime in rt
| -- Looking for clock_gettime in rt - not found
| -- Looking for sem_wait in rt
| -- Looking for sem_wait in rt - not found
| -- Looking for sem_wait in pthread
| -- Looking for sem_wait in pthread - found
| -- Looking for gethostbyname in nsl
| -- Looking for gethostbyname in nsl - not found
| -- Looking for socket in socket
| -- Looking for socket in socket - not found
| -- Looking for CLOCK_MONOTONIC
| -- Looking for CLOCK_MONOTONIC - not found
| -- Looking for atomic_cas_32
| -- Looking for atomic_cas_32 - not found
| -- Looking for AF_UNIX
| -- Looking for AF_UNIX - found
| -- Looking for backtrace_symbols_fd
| -- Looking for backtrace_symbols_fd - found
| -- Performing Test NN_HAVE_MSG_CONTROL
| -- Performing Test NN_HAVE_MSG_CONTROL - Success
| -- Performing Test NN_HAVE_GCC_ATOMIC_BUILTINS
| -- Performing Test NN_HAVE_GCC_ATOMIC_BUILTINS - Success
| -- Using asciidoctor at /opt/local/bin/asciidoctor
| -- Configuring done (16.1s)
| CMake Warning (dev) at src/CMakeLists.txt:340 (add_library):
| Policy CMP0156 is not set: De-duplicate libraries on link lines based on
| linker capabilities.  Run "cmake --help-policy CMP0156" for policy details.
| Use the cmake_policy command to set the policy and suppress this warning.

| Since the policy is not set, legacy libraries de-duplication strategy will
| be applied.
| This warning is for project developers.  Use -Wno-dev to suppress it.

| -- Generating done (0.2s)
| -- Build files have been written to: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanomsg/__CMake_build

nanomsg-cmake| CMake configuration: SUCCEEDED
nanomsg-cmake| CMake project nanomsg  has 74 build targets.
nanomsg-cmake| Generated Meson AST: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanomsg/meson.build
nanomsg-cmake| Project name: nanomsg
nanomsg-cmake| Project version: undefined
nanomsg-cmake| C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
nanomsg-cmake| C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
nanomsg-cmake| Build targets in project: 79
nanomsg-cmake| Subproject nanomsg-cmake finished.

Library anl found: NO
Library ws2_32 found: NO
Library mswsock found: NO
Run-time dependency nanopb found: NO (tried cmake)

Executing subproject nanopb-cmake method cmake 


| Configuring the build directory with CMake version 3.30.3
| Running CMake with: -G Ninja -DCMAKE_INSTALL_PREFIX=/opt/local -Dnanopb_BUILD_GENERATOR=False -DBUILD_SHARED_LIBS=False -DCMAKE_C_FLAGS=-DPB_ENABLE_MALLOC=1 -DPB_NO_PACKED_STRUCTS=1 -DCMAKE_POSITION_INDEPENDENT_CODE=True -DCMAKE_BUILD_TYPE=Debug
| - build directory:          /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanopb/__CMake_build
| - source directory:         /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2/subprojects/nanopb
| - toolchain file:           /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanopb/__CMake_build/CMakeMesonToolchainFile.cmake
| - preload file:             /opt/local/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/mesonbuild/cmake/data/preload.cmake
| - trace args:               --trace-expand --trace-format=json-v1 --no-warn-unused-cli --trace-redirect=cmake_trace.txt
| - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089, CMP0102]

| CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
| Compatibility with CMake < 3.5 will be removed from a future version of
| CMake.

| Update the VERSION argument <min> value or use a ...<max> suffix to tell
| CMake that the project does not need compatibility with older versions.

| Put cmake in trace mode, but with variables expanded.
| Put cmake in trace mode and sets the trace output format.
| Not searching for unused variables given on the command line.
| Put cmake in trace mode and redirect trace output to a file instead of stderr.
| Trace will be written to cmake_trace.txt

| -- The C compiler identification is GNU 4.2.1
| -- Checking whether C compiler has -isysroot
| -- Checking whether C compiler has -isysroot - yes
| -- Checking whether C compiler supports OSX deployment target flag
| -- Checking whether C compiler supports OSX deployment target flag - yes
| -- Detecting C compiler ABI info
| -- Detecting C compiler ABI info - done
| -- Check for working C compiler: /usr/bin/gcc-4.2 - skipped
| -- Detecting C compile features
| -- Detecting C compile features - done
| -- Found Python: /opt/local/bin/python3.12 (found version "3.12.6") found components: Interpreter
| -- Configuring done (3.9s)
| CMake Warning (dev) at CMakeLists.txt:112 (add_library):
| Policy CMP0156 is not set: De-duplicate libraries on link lines based on
| linker capabilities.  Run "cmake --help-policy CMP0156" for policy details.
| Use the cmake_policy command to set the policy and suppress this warning.

| Since the policy is not set, legacy libraries de-duplication strategy will
| be applied.
| This warning is for project developers.  Use -Wno-dev to suppress it.

| -- Generating done (0.1s)
| -- Build files have been written to: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanopb/__CMake_build

nanopb-cmake| CMake configuration: SUCCEEDED
nanopb-cmake| CMake project nanopb  has 1 build targets.
nanopb-cmake| Generated Meson AST: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/subprojects/nanopb/meson.build
nanopb-cmake| Project name: nanopb
nanopb-cmake| Project version: undefined
nanopb-cmake| C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
nanopb-cmake| C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
nanopb-cmake| Build targets in project: 80
nanopb-cmake| Subproject nanopb-cmake finished.

Run-time dependency libgit2 found: YES 1.7.2
Run-time dependency boxfort found: NO (tried pkgconfig, framework and cmake)
Looking for a fallback subproject for the dependency boxfort

Executing subproject boxfort 

boxfort| Project name: boxfort
boxfort| Project version: 0.0.1
boxfort| C compiler for the host machine: ccache /usr/bin/gcc-4.2 (gcc 4.2.1 "powerpc-apple-darwin10-gcc-4.2.1 (GCC) 4.2.1 (Apple Inc. build 5621) (dot 3)")
boxfort| C linker for the host machine: /usr/bin/gcc-4.2 ld64 87.3.1
boxfort| Program python3 found: YES
boxfort| Program git found: YES (/opt/local/bin/git)
boxfort| Compiler for C supports arguments -Wno-unused-parameter: YES (cached)
boxfort| Compiler for C supports arguments -Wno-unused-value: YES (cached)
boxfort| Compiler for C supports arguments -fvisibility=hidden: YES (cached)
boxfort| Compiler for C supports arguments /SAFESEH:NO: NO (cached)
boxfort| Compiler for C supports arguments /source-charset:utf-8: NO (cached)
boxfort| Dependency threads found: YES unknown (cached)
boxfort| Library rt found: NO
boxfort| Library m found: YES

subprojects/boxfort/meson.build:92:1: ERROR: Problem encountered: Architecture "ppc" is not supported.

A full log can be found at /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build/meson-logs/meson-log.txt
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2" && /opt/local/bin/meson setup --prefix=/opt/local /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/Criterion-2.4.2 /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_criterion/criterion/work/build --wrap-mode=default 
Exit code: 1
```

Thankfully, it is not needed now to build `syslog-ng`.

**Comment by HofiOne** [comment]
> @Schamschula I remembered now: I wanted to add `criterion` before (not sure for which dependent), but it is broken on powerpc due to boxfort:
> Thankfully, it is not needed now to build `syslog-ng`.

yeah, criterion dependency is fixed now, so turning off tests (`-DBUILD_TESTING=OFF`) will turn off its dependency as well

**Comment by barracuda156** [comment]
@HofiOne By the way will it be problematic to use something cross-platform for tests? There are a lot of test frameworks, and this is the only one which is broken, AFAIK.

**Comment by HofiOne** [comment]
> @HofiOne The port has no maintainer at the moment (would you like to be one?), just @Schamschula and myself independently decided to finally update it.

Yes, we could have multiple other useful cross-platform builds (e.g. for BSD and/or its variants), but 

- we have a bunch of CI tests already, and the builds run a bit too long already
- such platforms are not supported natively by GitHub Actions, so we would either have to setup external CI runners, or use some cross-platform solution
The latter is again a bunch of effort that we do not have resources actually, but again, any contributions are very welcome ;)

What I'd like to have back at least is the gcc builds for macOS as well (that was turned off earlier when the gcc drop was decided)

**Comment by barracuda156** [comment]
@HofiOne I did not mean to check anything on exotic platforms. What I meant is to drop an exotic `criterion` in favor of whatever standard testing framework fits requirements. This is literally the only software I can name which uses `criterion` for tests. We have ~40k in MacPorts.

**Comment by HofiOne** [comment]
Oh, I was careless, sorry.
There is a small chance we might switch over in the near future, as many tests are already using it. Although it's legacy, it currently fulfills its purpose. However, due to a lack of resources, this task has been pushed to the back of our to-do list.
